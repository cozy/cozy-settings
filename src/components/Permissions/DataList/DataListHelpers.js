export const sortPermissionsByName = (permissionsToDisplay, t) => {
  return Object.entries(permissionsToDisplay).sort((a, b) => {
    return t('CozyPermissions.Permissions.' + a[0])
      .toString()
      .localeCompare(t('CozyPermissions.Permissions.' + b[0]))
  })
}
