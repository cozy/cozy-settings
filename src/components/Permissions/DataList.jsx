import React from 'react'
import { Link } from 'react-router-dom'
import Page from 'components/Page'
import PageTitle from 'components/PageTitle'
import Typography from 'cozy-ui/transpiled/react/Typography'
import NavigationList, {
  NavigationListSection
} from 'cozy-ui/transpiled/react/NavigationList'
import ListItem from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItem'
import ListItemText from 'cozy-ui/transpiled/react/ListItemText'
import ListItemIcon, {
  mediumSize
} from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItemIcon'
import Divider from 'cozy-ui/transpiled/react/MuiCozyTheme/Divider'
import Icon from 'cozy-ui/transpiled/react/Icon'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import { APPS_DOCTYPE, KONNECTORS_DOCTYPE } from 'doctypes'
import withAllLocales from 'lib/withAllLocales'
import { getPermissionIconName } from 'containers/helpers/permissionsHelper'
import { completePermission, sortPermissionsByName } from './DataListHelpers'
import { routes } from 'constants/routes'
import CozyClient, {
  Q,
  useQuery,
  isQueryLoading,
  hasQueryBeenLoaded
} from 'cozy-client'

const DataList = ({ t }) => {
  const THIRTY_SECONDS = 30 * 1000
  const queryResultApps = useQuery(Q(APPS_DOCTYPE), {
    as: APPS_DOCTYPE,
    fetchPolicy: CozyClient.fetchPolicies.olderThan(THIRTY_SECONDS)
  })
  const queryResultKonnectors = useQuery(Q(KONNECTORS_DOCTYPE), {
    as: KONNECTORS_DOCTYPE,
    fetchPolicy: CozyClient.fetchPolicies.olderThan(THIRTY_SECONDS)
  })

  const permissionsToDisplay = completePermission(
    queryResultApps,
    queryResultKonnectors
  )

  return (
    <Page narrow>
      <PageTitle>{t('Permissions.data')}</PageTitle>
      {(isQueryLoading(queryResultApps) ||
        isQueryLoading(queryResultKonnectors)) &&
      (!hasQueryBeenLoaded(queryResultApps) ||
        !hasQueryBeenLoaded(queryResultKonnectors)) ? (
        <Spinner size="large" className="u-flex u-flex-justify-center u-mt-1" />
      ) : queryResultApps.fetchStatus === 'failed' ||
        queryResultKonnectors.fetchStatus === 'failed' ? (
        <Typography variant="body1" className="u-mb-1-half">
          {t('Permissions.failedRequest')}
        </Typography>
      ) : (
        <NavigationList>
          <NavigationListSection>
            {sortPermissionsByName(permissionsToDisplay, t).map(
              ([key, slugs]) => {
                const iconName = getPermissionIconName(key)
                return (
                  <React.Fragment key={key}>
                    <Link to={`${routes.dataList}/${key}`}>
                      <ListItem button>
                        <ListItemIcon>
                          <Icon
                            icon={
                              require(`cozy-ui/transpiled/react/Icons/${iconName}`)
                                .default
                            }
                            size={mediumSize}
                          />
                        </ListItemIcon>
                        <ListItemText
                          primary={t('CozyPermissions.Permissions.' + key)}
                          secondary={t('Permissions.numberOfApplications', {
                            smart_count: slugs.length
                          })}
                        />
                      </ListItem>
                    </Link>
                    <Divider />
                  </React.Fragment>
                )
              }
            )}
          </NavigationListSection>
        </NavigationList>
      )}
    </Page>
  )
}

export default withAllLocales(DataList)
