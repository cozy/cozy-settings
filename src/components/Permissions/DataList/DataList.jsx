import React from 'react'

import Typography from 'cozy-ui/transpiled/react/Typography'
import NavigationList, {
  NavigationListSection
} from 'cozy-ui/transpiled/react/NavigationList'
import Divider from 'cozy-ui/transpiled/react/Divider'
import ListItem from 'cozy-ui/transpiled/react/ListItem'
import ListItemText from 'cozy-ui/transpiled/react/ListItemText'
import ListItemIcon, {
  smallSize,
  mediumSize
} from 'cozy-ui/transpiled/react/ListItemIcon'
import Icon from 'cozy-ui/transpiled/react/Icon'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import ListItemSecondaryAction from 'cozy-ui/transpiled/react/ListItemSecondaryAction'
import RightIcon from 'cozy-ui/transpiled/react/Icons/Right'

import withAllLocales from 'lib/withAllLocales'
import { routes } from 'constants/routes'
import {
  completePermission,
  getPermissionIconName
} from 'components/Permissions/helpers/permissionsHelper'
import useAppsOrKonnectors from 'components/Permissions/hooks/useAppsOrKonnectors'
import { sortPermissionsByName } from 'components/Permissions/DataList/DataListHelpers'

const DataList = ({ t }) => {
  const { isResultLoading, hasQueryFailed, appsResult, konnectorsResult } =
    useAppsOrKonnectors()

  const permissionsToDisplay = completePermission(appsResult, konnectorsResult)

  return (
    <>
      {isResultLoading ? (
        <Spinner size="large" className="u-flex u-flex-justify-center u-mt-1" />
      ) : hasQueryFailed ? (
        <Typography variant="body1" className="u-mb-1-half">
          {t('Permissions.failedRequest')}
        </Typography>
      ) : (
        <NavigationList>
          <NavigationListSection>
            {sortPermissionsByName(permissionsToDisplay, t).map(
              ([key, slugs]) => {
                const iconName = getPermissionIconName(key)
                return (
                  <React.Fragment key={key}>
                    <ListItem
                      button
                      component="a"
                      href={`/#${routes.dataList}/${key}`}
                    >
                      <ListItemIcon>
                        <Icon
                          icon={
                            require(`cozy-ui/transpiled/react/Icons/${iconName}`)
                              .default
                          }
                          size={mediumSize}
                        />
                      </ListItemIcon>
                      <ListItemText
                        primary={t('CozyPermissions.' + key)}
                        secondary={t('Permissions.numberOfApplications', {
                          smart_count: slugs.length
                        })}
                      />
                      <ListItemSecondaryAction>
                        <Icon
                          icon={RightIcon}
                          size={smallSize}
                          className="u-mr-1"
                          style={{ color: 'var(--secondaryTextColor)' }}
                        />
                      </ListItemSecondaryAction>
                    </ListItem>
                    <Divider variant="inset" />
                  </React.Fragment>
                )
              }
            )}
          </NavigationListSection>
        </NavigationList>
      )}
    </>
  )
}

export default withAllLocales(DataList)
