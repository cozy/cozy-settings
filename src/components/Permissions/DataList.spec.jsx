import { render } from '@testing-library/react'
import React from 'react'
import DataList from './DataList'
import AppLike from 'test/AppLike'

jest.mock('components/Page', () => {
  // eslint-disable-next-line react/display-name
  return ({ narrow, children }) => (
    <div data-testid="page" data-narrow={narrow}>
      {children}
    </div>
  )
})

jest.mock('components/PageTitle', () => {
  // eslint-disable-next-line react/display-name
  return ({ children }) => <div data-testid="page-title">{children}</div>
})

describe('DataList', () => {
  it('should display DataList title', () => {
    const { getByText } = render(
      <AppLike>
        <DataList />
      </AppLike>
    )
    expect(getByText('Data')).toBeTruthy()
  })
})
