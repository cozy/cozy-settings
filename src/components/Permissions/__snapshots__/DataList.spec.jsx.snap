// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataList should display DataList title 1`] = `
<div>
  <div
    data-narrow="true"
    data-testid="page"
  >
    <div
      data-testid="page-title"
    >
      Data
    </div>
  </div>
</div>
`;
